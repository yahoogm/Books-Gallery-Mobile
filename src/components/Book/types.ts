import {VolumeInfo} from '../../types/types';

type DetailBookType = {
  volumeInfo: VolumeInfo;
};

export type BookProps = {
  uriDetailBook: string;
  authorsName: string[] | string;
  categories: string[] | string;
  detailBook: DetailBookType;
  setShowBook: React.Dispatch<React.SetStateAction<boolean>>;
  refReadBook: React.RefObject<any>;
};

export type BookContainerProps = {
  setShowBook: React.Dispatch<React.SetStateAction<boolean>>;
  refReadBook: React.RefObject<any>;
};
