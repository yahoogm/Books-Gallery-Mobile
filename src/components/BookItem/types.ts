export type BookItemProps = {
  detailBook: string | number | string[];
  label: string;
};
