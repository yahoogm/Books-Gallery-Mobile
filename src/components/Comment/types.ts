import {NativeStackNavigationProp} from '@react-navigation/native-stack';
import {BookItem, Comment, RootStackParamList} from '../../types/types';
import {ObjectSchema} from 'yup';

export type FormCommentValues = {
  comment: string;
};

export type AddCommentValues = {
  userName: string;
  bookId: string;
  profilePic: string;
  id: string;
  ulasan: string;
  userId: string;
  createdAt: string;
  updatedAt: string;
};

// type for handling navigation to login page
export type NavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  'Login'
>;

export type ToastRenderProps = {
  id: string;
};

export type UserItem = {
  email: string;
  familyName: string;
  givenName: string;
  id: string;
  name: string;
  photo: string;
};

export type HandleBookCommentFunction = {
  handleNavigateUserNotLogin: () => void;
  handleAddCommentBook: (review: AddCommentValues) => void;
  handleDeleteCommentBook: () => void;
  handleUpdateCommentBook: (ulasan: string, updateAt: string) => void;
  getIdDocumentFromFirestore: (id: string) => void;
};

export type CommentProps = {
  refOpenDeleteModal: React.RefObject<any>;
  deleteModal: boolean;
  setDeleteModal: React.Dispatch<React.SetStateAction<boolean>>;
  handleButton: HandleBookCommentFunction;
  isEdit: boolean;
  setIsEdit: React.Dispatch<React.SetStateAction<boolean>>;
  commentSchema: ObjectSchema<FormCommentValues>;
  initialValues: FormCommentValues;
  isLogin: boolean;
  user: UserItem;
  detailBook: BookItem;
  commentBook: Array<Comment>;
};
