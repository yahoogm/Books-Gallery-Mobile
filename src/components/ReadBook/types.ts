export type ReadBookProps = {
  setShowBook: React.Dispatch<React.SetStateAction<boolean>>;
  showBook: boolean;
  refReadBook: React.RefObject<any>;
  isbn: number;
};
