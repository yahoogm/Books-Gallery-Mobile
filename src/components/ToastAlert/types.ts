export type ToastAlertProps = {
  id: string;
  message: string;
  action: 'success' | 'info' | 'error' | 'warning' | 'attention';
};
