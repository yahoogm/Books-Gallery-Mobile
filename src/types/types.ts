type VolumeInfo = {
  title: string;
  publishedDate: string;
  description: string;
  authors: Array<string>;
  publisher: string;
  pageCount: number;
  imageLinks: {smallThumbnail: string; thumbnail: string; large?: string};
  categories: Array<string>;
  industryIdentifiers: Array<{identifier: string; type: string}>;
};

export type Review = {
  bookId: string;
  createdAt: {};
  id: string;
  profilePic: string;
  ulasan: string;
  updatedAt: {};
  userId: string;
  userName: string;
};

export type BookItem = {
  id: string;
  volumeInfo: VolumeInfo;
};

export type BookState = {
  type: string;
  reviewId: string;
  detailBook: BookItem;
  books: {
    items: [];
  };
  reviewBook: Array<Review>;
};

export type userItem = {
  email: string;
  familyName: string;
  givenName: string;
  id: string;
  name: string;
  photo: string;
};

export type RootStackParamList = {
  DetailBook: {bookId: string};
  Login: undefined;
};
