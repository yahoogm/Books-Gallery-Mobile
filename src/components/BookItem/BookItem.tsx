import {Heading, Text, View} from '@gluestack-ui/themed';
import {BookItemProps} from './types';
const BookItem: React.FC<BookItemProps> = ({detailBook, label}) => {
  return (
    <View alignItems="center">
      <Heading size="md">{label}:</Heading>
      <Text>{detailBook}</Text>
    </View>
  );
};

export default BookItem;
