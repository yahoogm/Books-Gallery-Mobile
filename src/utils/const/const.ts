export const modifiedName = (names: string[] | undefined) => {
  if (!Array.isArray(names) || names.length === 0) {
    return '';
  }

  if (names == undefined) {
    return '';
  }

  const nonEmptyNames = names.filter(name => name.trim() !== '');

  if (nonEmptyNames.length === 0) {
    return '';
  }

  return nonEmptyNames.join(', ');
};
