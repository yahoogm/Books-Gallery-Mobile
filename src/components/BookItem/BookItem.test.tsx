import BookItem from './BookItem';
import {BookItemProps} from './types';
import {modifiedName} from '../../utils/const/const';
import {render} from '@testing-library/react-native';

describe('BookItem component', () => {
  const rendererComponent = (props: BookItemProps) =>
    render(<BookItem {...props} />);

  //   snapshoot test
  it('match snapshot with string detail book', () => {
    const props: BookItemProps = {
      detailBook: 'this book is very nice',
      label: 'Description',
    };
    const tree = rendererComponent(props).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot with number detail book', () => {
    const props: BookItemProps = {
      detailBook: 123456,
      label: 'Page Count',
    };

    const tree = rendererComponent(props).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot with string in array detail book', () => {
    const props: BookItemProps = {
      detailBook: ['budi', 'tono'],
      label: 'Publisher',
    };

    const tree = rendererComponent(props).toJSON();
    expect(tree).toMatchSnapshot();
  });

  //   assertion test
  it('renders correctly with string detail book', () => {
    const props: BookItemProps = {
      detailBook: 'this book is very nice',
      label: 'Description',
    };

    const component = rendererComponent(props);
    const instance = component.root;

    expect(instance.findByProps({children: 'Description:'})).toBeTruthy();
    expect(
      instance.findByProps({children: 'this book is very nice'}),
    ).toBeTruthy();
  });

  it('renders correctly with number detail book', () => {
    const props: BookItemProps = {
      detailBook: 1234567,
      label: 'Page Count',
    };

    const component = rendererComponent(props);
    const instance = component.root;

    expect(instance.findByProps({children: 'Page Count:'})).toBeTruthy();
    expect(instance.findByProps({children: 1234567})).toBeTruthy();
  });

  it('renders correctly with string in array detail book', () => {
    const items = modifiedName(['budi', 'tono']);
    const props: BookItemProps = {
      detailBook: items,
      label: 'Publisher',
    };

    const component = rendererComponent(props);
    const instance = component.root;

    expect(instance.findByProps({children: 'Publisher:'})).toBeTruthy();
    expect(instance.findByProps({children: 'budi, tono'})).toBeTruthy();
  });
});
