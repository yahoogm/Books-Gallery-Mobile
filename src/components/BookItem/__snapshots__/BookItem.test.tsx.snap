// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookItem component match snapshot with number detail book 1`] = `
<View
  data-style=""
  dataSet={
    {
      "style": "",
    }
  }
  style={
    [
      undefined,
      {},
      {
        "alignItems": "center",
      },
      undefined,
    ]
  }
>
  <Text
    accessibilityRole="header"
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    size="md"
    style={
      {
        "fontSize": 32,
        "fontWeight": "bold",
        "marginVertical": 42.56,
      }
    }
  >
    Page Count:
  </Text>
  <Text
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        undefined,
      ]
    }
  >
    123456
  </Text>
</View>
`;

exports[`BookItem component match snapshot with string detail book 1`] = `
<View
  data-style=""
  dataSet={
    {
      "style": "",
    }
  }
  style={
    [
      undefined,
      {},
      {
        "alignItems": "center",
      },
      undefined,
    ]
  }
>
  <Text
    accessibilityRole="header"
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    size="md"
    style={
      {
        "fontSize": 32,
        "fontWeight": "bold",
        "marginVertical": 42.56,
      }
    }
  >
    Description:
  </Text>
  <Text
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        undefined,
      ]
    }
  >
    this book is very nice
  </Text>
</View>
`;

exports[`BookItem component match snapshot with string in array detail book 1`] = `
<View
  data-style=""
  dataSet={
    {
      "style": "",
    }
  }
  style={
    [
      undefined,
      {},
      {
        "alignItems": "center",
      },
      undefined,
    ]
  }
>
  <Text
    accessibilityRole="header"
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    size="md"
    style={
      {
        "fontSize": 32,
        "fontWeight": "bold",
        "marginVertical": 42.56,
      }
    }
  >
    Publisher:
  </Text>
  <Text
    data-style=""
    dataSet={
      {
        "style": "",
      }
    }
    style={
      [
        undefined,
        {},
        undefined,
      ]
    }
  >
    budi
    tono
  </Text>
</View>
`;
